python/cuda_parallel/cuda/parallel/experimental/__init__.py

```diff
@@ -164,30 +145,54 @@ def handle(self):
         return _CCCLOp(_CCCLOpKindEnum.STATELESS, self.name, ctypes.c_char_p(self.ltoir), len(self.ltoir), 1, 1, None)
 
 
-def _extract_ctypes_ltoirs(numba_cuda_compile_results):
-    view_lst = [_CCCLStringView(ltoir, len(ltoir))
-                for ltoir in numba_cuda_compile_results]
-    view_arr = (_CCCLStringView * len(view_lst))(*view_lst)
-    return ctypes.pointer(_CCCLStringViews(view_arr, len(view_arr)))
-
-
-def _facade_iter_as_cccl_iter(d_in):
-    def prefix_name(name):
-        return (d_in.prefix + "_" + name).encode('utf-8')
-    # type name ltoi ltoir_size size alignment state
-    adv = _CCCLOp(_CCCLOpKindEnum.STATELESS, prefix_name("advance"), None, 0, 1, 1, None)
-    drf = _CCCLOp(_CCCLOpKindEnum.STATELESS, prefix_name("dereference"), None, 0, 1, 1, None)
-    info = _numba_type_to_info(d_in.ntype)
-    ltoirs = _extract_ctypes_ltoirs(d_in.ltoirs)
-    # size alignment type advance dereference value_type state ltoirs
-    return _CCCLIterator(d_in.size, d_in.alignment, _CCCLIteratorKindEnum.ITERATOR, adv, drf, info, d_in.state_c_void_p, ltoirs)
+def _device_array_to_cccl_iter(array):
+    dtype = array.dtype
+    info = _numpy_type_to_info(dtype)
+    # Note: this is slightly slower, but supports all ndarray-like objects as long as they support CAI
+    # TODO: switch to use gpumemoryview once it's ready
+    return _CCCLIterator(1, 1, _CCCLIteratorKindEnum.POINTER, _CCCLOp(), _CCCLOp(), info, array.__cuda_array_interface__["data"][0])
 
 
-def _d_in_as_cccl_iter(d_in):
-    if hasattr(d_in, 'ntype'):
-        return _facade_iter_as_cccl_iter(d_in)
-    assert hasattr(d_in, 'dtype')
-    return _device_array_to_pointer(d_in)
+def _iterator_to_cccl_iter(it):
+    context = cuda.descriptor.cuda_target.target_context
+    numba_type = it._numba_type
+    size = context.get_value_type(numba_type).get_abi_size(context.target_data)
+    alignment = context.get_value_type(
+        numba_type).get_abi_alignment(context.target_data)
+    advance_ltoir, deref_ltoir = it.ltoirs
+    advance_op = _CCCLOp(
+        _CCCLOpKindEnum.STATELESS,
+        type(it).advance.__name__.encode("utf-8"),
```

Comment:
I'd like to understand the requirements from the op's `name` here. Do the names need to be globally unique? Unique within the context of a single NVRTC compilation? Something else?

Reply:
xref: https://github.com/NVIDIA/cccl/pull/3118#discussion_r1879351174